### PR TITLE
Support looking up missing symbols during compilation

### PR DESCRIPTION
**Pull request by pyrmont**
This PR supports looking up missing symbols during the compilation phase of evaluation by calling a function stored under the dynamic binding `:missing-symbol`.

## Background

Using the recently added `:redef` dynamic binding, the following workflow is now possible in a REPL-connected editor:[^1]

1. evaluate a buffer that imports a Janet module
2. in a separate buffer, edit the function definition contained in the Janet module
3. evaluate just the function definition
4. use the redefined function in the original buffer without further action

This is really cool and I'm extremely glad it's possible. Of course I'd like to do more. The following is not currently possible:

1. evaluate a buffer that imports a Janet module
2. _add a function definition_ to the Janet module
3. evaluate the function definition
4. use the new function in the original buffer without further action

## Implementation

This PR does two things:

1. If `janet_resolve_ext()` does not resolve a symbol in the environment, `janet_resolve_ext()` now checks if there is a function stored under the dynamic binding `:missing-symbol`. If there is, it calls that function with the symbol and the environment.

2. When a module is imported, the environment is added to a table keyed by prefix. The table is stored under the `:modules` dynamic binding.

These changes make the above-desired workflow possible. The user would provide a missing symbol lookup function that extracts the prefix from a missing symbol, checks whether a module in the current environment exists under that prefix and, if there is, checks whether the unprefixed symbol exists in the module's environment and, if so, returns its value.

Since the additional check in `janet_resolve_ext()` only occurs during compilation and only occurs when a symbol is not resolved, it adds minimal overhead. 

[^1]: I'm doing this in Neovim using the [Conjure plugin](https://github.com/Olical/conjure) and a custom netrepl server I'm running using my in-development project management tool, [Jeep](https://github.com/pyrmont/jeep). The netrepl server includes a custom `run-context` function that uses separate environments.

### CODE REVIEW

**Comment by bakpakin** [comment]
This is pretty interesting, I was wondering if there are any bad implications here, but aside from the fact that the `:missing-symbol` handler here needs to resolve synchronously, I can't think of any. However, that would be an issue if a user tried to load a symbol from over a network or pipe, limiting the utility of this.

Perhaps better would be to somehow retry the compilation if you get a compiler error with "unknown symbol <symbol name>", which would allow more sophisticated recovery that could look up symbol asynchronously. That said, it might be better to force things to be synchronous for compilation, so I'm not entirely convinced that this is really much of an issue.

File: src/boot/boot.janet [hunk 1]
```diff
@@ -2746,6 +2746,8 @@
                 (and as (string as "/"))
                 prefix
                 (string (last (string/split "/" path)) "/")))
+  (unless (zero? (length prefix))
+    (put-in env [:modules prefix] newenv))
```
Comment (bakpakin) [inline src/boot/boot.janet:2750]: Why is this here? Is this a local change that got added by accident? This seems like a local hack you added for your repl use-case.

File: src/core/util.c [hunk 2]
```diff
@@ -607,8 +607,14 @@ JanetBinding janet_resolve_ext(JanetTable *env, const uint8_t *sym) {
...
-    if (!janet_checktype(entry, JANET_TABLE))
-        return binding;
+    if (!janet_checktype(entry, JANET_TABLE)) {
+        Janet fn_entry = janet_table_get(env, janet_ckeywordv("missing-symbol"));
+        if (janet_checktype(fn_entry, JANET_FUNCTION)) {
+            Janet args[2] = { janet_wrap_symbol(sym), janet_wrap_table(env) };
+            entry = janet_call(janet_unwrap_function(fn_entry), 2, args);
+        }
+        if (!janet_checktype(entry, JANET_TABLE)) return binding;
+    }
```
Comment (bakpakin) [inline src/core/util.c:617]: So another issue with using janet_call is that currently, janet_compile will not panic. This is by design so that we can always easily get compilation errors. Instead, do what macro expansion does and use janet_pcall.

**Review by bakpakin — CHANGES_REQUESTED** [review]
A few changes need to be made for me to accept this, although I'm not totally against the idea.

File: src/boot/boot.janet (hunk 1 above)
Comment (pyrmont) [reply src/boot/boot.janet]: I'm sorry; I neglected to explain this in the PR write-up. I thought it would be useful generally to have the environments for modules stored in a dynamic binding in each environment. However, that's a change that's orthogonal to the goal of this PR and so I've removed.

File: src/core/util.c (hunk 2 above)
Comment (pyrmont) [reply src/core/util.c:617]: If I'm following the code correctly, [`macroexpand1`](https://github.com/janet-lang/janet/blob/f2f421a0a25aa2fd850372d9c160eac6bb3b8fe3/src/core/compile.c#L664-L699) creates a fiber to execute the function manually. I just want to confirm, would you prefer me to do it that way or is calling `janet_pcall` fine?

Comment (bakpakin) [reply src/core/util.c:617]: Ah, yeah they are quite similar but creating the fiber manually is fine. Just we shouldn't panic and instead issue a normal compilation error. Perhaps the compiler could be updated to use the usual panic path, but for now we should instead report errors with janetc_error.

Comment (pyrmont) [reply src/core/util.c]: I think the way that I'd originally implemented this wasn't correct. I put the lookup in `janet_resolve_ext()` rather than in `janetc_resolve()`. That didn't allow for the compiler to look up the symbol and output appropriate errors if there was a problem with the lookup handler.

To fix that, I've moved the missing symbol lookup into `compile.c` and added errors analogous to those that are output during macro expansion. Let me know if this isn't what you had in mind.

Comment (bakpakin) [reply src/core/util.c]: That looks good so far, yes. Essentially we just need to make sure that the `compile` function always outputs good errors. If it doesn't, you can get stuff that is _really_ hard to debug.

**Comment by bakpakin** [comment]
Ok, looks good to me.

**Comment by harryvederci** [comment]
This sounds really interesting!

> I'm doing this in Neovim using the Conjure plugin and a custom netrepl server I'm running using my in-development project management tool, Jeep. The netrepl server includes a custom run-context function that uses separate environments.

@pyrmont  Could you share this in a gist or something?

**Comment by pyrmont** [comment]
@harryvederci The custom function is [here](https://github.com/pyrmont/jeep/blob/1219041aeba7939ca20b2d39b6272e7f041b81c9/jeep/netrepl.janet#L18-L174). For further discussion, I'd be happy to discuss in Jeep's [Issues](https://github.com/pyrmont/jeep/issues) section :)
